### PR TITLE
#A20. As Admin, I would like to be able to administrate users better; add or remove participants and judges from teams and hackathons

### PR DESCRIPTION
**Pull request by stefdworschak**
## Description
- Adding new permissions approach with roles and groups setting up more detailed permissions
- Adding new views and extra features to HackAdmin panel
- Removing no longer needed `submissions` django app

## Pull request type

- [ ] #M01 (Miscellaneous User Story #01)
- [ ] #P02 (Participant User Story #02)
- [ ] #S03 (Staff User Story #03)
- [x] #A04 (Admin User Story #04)
- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## Related Issue
#183

## Configuration instructions
- Run migrations
- Load `groups.json` as fixture

## Testing
Tests are passing.

## Screenshots
image
image
image

### CODE REVIEW

**Review by TravelTimN — COMMENTED** [review]
A few things mentioned below.... suggestions / possible improvements, etc.

I started mentioning the code-blocks like these:
`{% if request.user.user_type|is_types:"SUPERUSER,STAFF,FACILITATOR_ADMIN,FACILITATOR_JUDGE,PARTNER_ADMIN,PARTNER_JUDGE" %}`

But the more I got into viewing the other files, the more I saw this was re-used in multiple places.
Would it not be better to define set groups for varied permissions, elsewhere?

If group A can do FOO, and group B can do BAR, why are we setting these within the different if-statements. Why not use something like a {% with %} block to set the varied groups apart, and just re-use the new "with-variable" being set.

`{% with group_A="FOO1, FOO2, FOO3, etc" %}{% endwith %}`
`{% with group_B="BAR1, BAR2, BAR3,etc" %}{% endwith %}`

Then these can be re-used as such:
`{% if user_type in group_A %}`
for example.... just a thought to clean it up a bit, but not necessary.

Apart from those suggestions, feel free to accept this as **LGTM**, and let me know, I'll adjust from Review-Comment to Approved

File: accounts/admin.py [hunk 1]
```diff
@@ -38,11 +38,14 @@ class CustomUserAdmin(BaseUserAdmin):
...
-    search_fields = ('full_name', 'email')
+    list_display = ('email', 'full_name', 'is_superuser', 'user_type',
+                    'is_external')
+    list_filter = ('is_staff', 'is_superuser', 'is_active', 'groups',
+                   'is_external')
+    search_fields = ('full_name', 'email', 'slack_display_name')
     ordering = ('email',)
-    readonly_fields = ('last_login', 'date_joined',)
+
+    readonly_fields = ('last_login', 'date_joined', 'user_type',)
```
[the remark this answers is not in the document]
Comment (stefdworschak) [reply accounts/admin.py:48]: It's to do with being a tuple, but I think we can remove it here.

File: accounts/models.py [hunk 2]
```diff
@@ -1,9 +1,23 @@
...
 from .lists import USER_TYPES_CHOICES, LMS_MODULES_CHOICES
 from teams.lists import LMS_LEVELS
 
+class UserType(Enum):
+    SUPERUSER = 0
+    STAFF = 1
+    FACILITATOR_ADMIN = 2
+    FACILITATOR_JUDGE = 3
+    FACILITATOR = 4
+    STUDENT = 5
```
[the remark this answers is not in the document]
Comment (stefdworschak) [reply accounts/models.py:15]: Agreed. Renaming this to "PARTICIPANT"

File: hackadmin/templates/all_users.html [hunk 3]
```diff
@@ -1,7 +1,11 @@
 {% extends 'base.html' %}
+{% load hackadmin_tags %}
+{% load account_tags %}
 
 {% block content %}
 
+{% include 'includes/back_button.html' with redirect_url="/hackadmin/" button_label="Back To HackAdmin" %}
```
[the remark this answers is not in the document]
Comment (stefdworschak) [reply hackadmin/templates/all_users.html:7]: Yeah, I found that out recently and find it really helpful.

File: static/js/script.js [hunk 4]
```diff
@@ -1,6 +1,30 @@
...
+        let judgeId = $(this).data('judge-id');
+        $('.judge_id').val(judgeId);
+    })
+
+    $('#add_participant_hackathon_id').change(function(){
+        let hackathonId = $(this).val();
+        $(`.hackadmin-team-select`).hide();
+        $(`#hackadmin-team-select-${hackathonId}`).show();
+    })
+
```
[the remark this answers is not in the document]
Comment (stefdworschak) [reply static/js/script.js:27]: Added and added.

File: profiles/views.py [hunk 5]
```diff
@@ -3,7 +3,7 @@
 from django.contrib.auth.decorators import login_required
 from django.conf import settings
 
-from accounts.models import CustomUser
+from accounts.models import CustomUser,UserType
```
[the remark this answers is not in the document]
Comment (stefdworschak) [reply profiles/views.py:6]: Added spacing

File: hackathon/views.py [hunk 6]
```diff
@@ -33,28 +31,44 @@
...
 
 
-class HackathonListView(LoginRequiredMixin, ListView):
-    """Renders a page with a list of Hackathons."""
-    model = Hackathon
-    ordering = ["-created"]
-    paginate_by = 8
+@login_required
+def list_hackathons(request):
+    """ Lists all hackathons available to a given student
```
[the remark this answers is not in the document]
Comment (stefdworschak) [reply hackathon/views.py:36]: Yes. This should be visible to anybody. We have in-template restrictions based on the UserType.

File: hackadmin/templates/list_all_users.html [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+                                <td>
+                                    {{ user.current_lms_module|readable_lms_module|title }}
+                                </td>
+                                <td>
+                                    {{user.user_type|readable_user_type}}
+                                </td>
+                                <td>
+                                    <a href="/profile/{{user.id}}/" class="ci-link mr-2" title="View Hackathon Profile"><i class="fas fa-eye"></i></a>
+                                    {% if slack_url and not user.is_external %}
+                                    {% split_string string=user.username delimiter="_" index=0  as slack_username %}
```
[the remark this answers is not in the document]
Comment (stefdworschak) [reply hackadmin/templates/list_all_users.html:49]: The `username` is different to the `slack_display_name`. The `username` is the user's ID and Workspace ID separated by an underscore. Using the split I am retrieving just the user's ID to generate a link to the user's Slack profile.

File: hackadmin/templates/includes/list_mentors.html [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+            <td>
+                {{ mentor_details.team.display_name }}
+            </td>
+            <td>
+                {{mentor_details.mentor.user_type|readable_user_type}}
+            </td>
+            <td>
+                <a href="/profile/{{mentor_details.mentor.id}}/" class="ci-link mr-2" title="View Hackathon Profile"><i class="fas fa-eye"></i></a>
+                {% if slack_url and not mentor_details.mentor.is_external %}
+                {% split_string string=mentor_details.mentor.username delimiter="_" index=0  as slack_username %}
```
[the remark this answers is not in the document]
Comment (stefdworschak) [reply hackadmin/templates/includes/list_mentors.html:32]: The `username` is different to the `slack_display_name`. The `username` is the user's ID and Workspace ID separated by an underscore. Using the split I am retrieving just the user's ID to generate a link to the user's Slack profile.

[the remark this answers is not in the document]
Comment (stefdworschak) [reply hackadmin/templates/includes/list_mentors.html:32]: Same as before.

File: hackadmin/templates/includes/list_judges.html [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+            <td>
+                {{ judge.current_lms_module }}
+            </td>
+            <td>
+                {{judge.user_type|readable_user_type}}
+            </td>
+            <td>
+                <a href="/profile/{{judge.id}}/" class="ci-link mr-2" title="View Hackathon Profile"><i class="fas fa-eye"></i></a>
+                {% if slack_url and not judge.is_external %}
+                {% split_string string=judge.username delimiter="_" index=0  as slack_username %}
```
[the remark this answers is not in the document]
Comment (stefdworschak) [reply hackadmin/templates/includes/list_judges.html:32]: Same as before.
The `username` is different to the `slack_display_name`. The `username` is the user's ID and Workspace ID separated by an underscore. Using the split I am retrieving just the user's ID to generate a link to the user's Slack profile.

File: hackadmin/templates/includes/list_participants.html [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+            <td>
+                {{ participant.current_lms_module|readable_lms_module|title }}
+            </td>
+            <td>
+                {{participant.user_type|readable_user_type}}
+            </td>
+            <td>
+                <a href="/profile/{{participant.id}}/" class="ci-link mr-2" title="View Hackathon Profile"><i class="fas fa-eye"></i></a>
+                {% if slack_url and not participant.is_external %}
+                {% split_string string=participant.username delimiter="_" index=0  as slack_username %}
```
[the remark this answers is not in the document]
Comment (stefdworschak) [reply hackadmin/templates/includes/list_participants.html:32]: Same as before.
The `username` is different to the `slack_display_name`. The `username` is the user's ID and Workspace ID separated by an underscore. Using the split I am retrieving just the user's ID to generate a link to the user's Slack profile.

File: profiles/templates/profiles/profile.html [hunk 11]
```diff
@@ -54,7 +55,7 @@
                   </div>
                   {% if '/accounts/edit_profile/' == request.path %}
                   {% else %}
-                    {% if is_owner or request.user.is_staff or user.profile_is_public %}
+                    {% if is_owner or request.user.user_type|is_types:"SUPERUSER,STAFF,FACILITATOR_ADMIN,FACILITATOR_JUDGE,PARTNER_ADMIN,PARTNER_JUDGE" or user.profile_is_public %}
```
[the remark this answers is not in the document]
Comment (stefdworschak) [reply profiles/templates/profiles/profile.html]: Makes sense. Changed that now.

**Comment by stefdworschak** [comment]
@TravelTimN changed it to use groups in `{% with %}` statements as you suggested. Could you please review again and approve if you are okay with the changes?

**Review by TravelTimN — APPROVED** [review]
LGTM! 👍🏻

**Pull request merged**